enhancements/results-forwarding.md

```diff
@@ -0,0 +1,379 @@
+
+* Provide a generic way to implement compliance result forwarding for Compliance Operator.
+  This should be flexible enough to accommodate the current implementation (CRDs and PVCs),
+  while allowing for custom forwarder implementations.
+
+* Implement and define a stable API for forwarder implementations.
+
+* Implement result/remediation CRDs and ARF PVC as reference implementations of the forwarding
+  model.
+
+### Non-Goals
+
+* Deprecate the current CRD/PVC approach: There is still a use-case for this.
+
+* Expand the "evidence" that's currently gathered by the Compliance Operator. This is
+  needed but should be done as part of a separate initiative.
+
+## Proposal
+
+In short, the proposal is to have Compliance Operator *always* forward Compliance results (suggested
+remediations being part of this), as well as evidence (e.g. ARF results). This will require
+changing the **aggregator** to always forward via a GRPC to a configured implementation. The
+**resultserver** then becomes subject to the evidence forwarding provider and will be renamed
+**evidence-persistor**.
+
+Forwarder implementations would be configured via the `ScanSetting` and they'd be called "providers".
+For backwards compatibility, we'd have defaults that would point to the current mode of operation
+of the Compliance Operator. e.g. a `pvc` provider for evidence storage.
+
+For more detail on the proposed API, see the *API Extensions* section.
+
+### User Stories
+
+* As an Site Reliability Engineer managing multiple clusters, I'd like to run compliance
+  scans on my fleet in such a way that uses the minimal amount of resources.
+
+* As a Site Reliability Engineer managing multiple clusters, I'd like to have
+  one place to view the compliance stance of my whole fleet.
+
+### API Extensions
+
+Given the scope of the change, we'd upgrade the ScanSetting's version to `v1alpha2`.
+
+A sample `ScanSetting` object with Compliance Operator's current mode of operation would look as follows:
+
+```yaml
+apiVersion: compliance.openshift.io/v1alpha2
+kind: ScanSetting
+metadata:
+  name: default
+  namespace: openshift-compliance
+generateResultCRDs: true  # Adds CRD creator and sets forwarder config to send info there
+resultForwarding:
+  ... # Potential configuration goes here.
+evidenceForwarding:
+  provider: pvc
+  pvc: # configuration goes here. This would be the same as `rawResultStorage` in ScanSetting v1alpha1
+    pvAccessModes:
+      - ReadWriteOnce 
+    rotation: 3 
+    size: 1Gi
+roles:
+- worker 
+- master 
+scanTolerations: 
+  default:
+  - operator: Exists
+  schedule: 0 1 * * * 
+```
+
+To forward results, we'd implement a `grpc` provider. A sample would look as follows:
+
+```yaml
+apiVersion: compliance.openshift.io/v1alpha2
+kind: ScanSetting
+metadata:
+  name: default
+  namespace: openshift-compliance
+generateResultCRDs: false # no CRD-creator container will be provisioned
+storeEvidenceInPVC: false # no PVC store container will be provisioned
```

Comment:
that would be a misconfiguration. the GRPCconfiguration would anyway be created by default to point to the PVC and CRD handlers... if you configure the GRPC configuration by yourself the results get forwarded to wherever you point them to. No magic there.

Reply:
I'm probably over-simplifying things, but it would be nice to have everything encapsulated into the `provider` configuration.

I could see a user setting up the `grpc` provider and then wondering why they have results in physical volumes because they didn't know they also needed to set `storeEvidenceInPVC=false`.

What if the forwarders accepted a list?

I could have a configuration for storing everything in PVCs (like what we do today):

```
evidenceForwarding:
  - pvc:
      pvAccessMode:
        - ReadWriteOnce
      rotation: 3
      size: 1Gi
resultForwarding:
  - pvc:
      pvAccessMode:
        - ReadWriteOnce
      rotation: 3
      size: 1Gi
```

I could have a configuration for storing evidence in PVCs (because they're huge files still), and everything else in an external service.

```
evidenceForwarding:
  - pvc:
      pvAccessMode:
        - ReadWriteOnce
      rotation: 3
      size: 1Gi
resultForwarding:
  - grpc:
      version: v1
      endpoint: $endpoint
      tls: $certificate
```

Or, I could setup the providers to write it everywhere all the time (gRPC and PVCs):

```
evidenceForwarding:
  - pvc:
      pvAccessMode:
        - ReadWriteOnce
      rotation: 3
      size: 1Gi
resultForwarding:
  - grpc:
      version: v1
      endpoint: $endpoint
      tls: $certificate
  - pvc:
      pvAccessMode:
        - ReadWriteOnce
      rotation: 3
      size: 1Gi
```

Reply:
I like your proposal, we could use that instead!